src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -0,0 +1,158 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include "tcp_sockcm_ep.h"
+#include <ucs/sys/sock.h>
+#include <ucs/async/async.h>
+#include <ucs/arch/bitops.h>
+#include <ucs/sys/string.h>
+
+
+static UCS_F_ALWAYS_INLINE
+uct_tcp_sockcm_t *uct_tcp_sockcm_ep_get_cm(uct_tcp_sockcm_ep_t *cep)
+{
+    /* return the tcp sockcm connection manager this ep is using */
+    return ucs_container_of(cep->super.super.super.iface, uct_tcp_sockcm_t,
+                            super.iface);
+}
+
+ucs_status_t uct_tcp_sockcm_ep_disconnect(uct_ep_h ep, unsigned flags)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_server_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+static ucs_status_t uct_tcp_sockcm_ep_client_init(uct_tcp_sockcm_ep_t *cep,
+                                                  const uct_ep_params_t *params)
+{
+    uct_tcp_sockcm_t *tcp_sockcm = uct_tcp_sockcm_ep_get_cm(cep);
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr *server_addr;
+    ucs_async_context_t *async_ctx;
+    uct_tcp_sa_arg_t *sa_arg_ctx;
+    ucs_status_t status;
+
+    cep->state |= UCT_TCP_SOCKCM_EP_ON_CLIENT;
+
+    server_addr = params->sockaddr->addr;
+    status = ucs_socket_create(server_addr->sa_family, SOCK_STREAM, &cep->fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the fd to non-blocking mode. (so that connect() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(cep->fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    /* try to connect to the server */
+    status = ucs_socket_connect(cep->fd, server_addr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        goto err_close_socket;
+    }
+    ucs_assert((status == UCS_OK) || (status == UCS_INPROGRESS));
```

Comment:
seems like code duplication with `UCS_CLASS_INIT_FUNC(uct_sockcm_ep_t`
can you pls extract common code (including ucs_async_set_event_handler)